File: InterviewBit/Arrays/Min_Steps_in_Infinite_Grid/min_steps_in_infinite_grid.rb
# https://www.interviewbit.com/problems/min-steps-in-infinite-grid

class Solution
  # @param a : array of integers
  # @param b : array of integers
  # @return an integer
  def coverPoints(a, b)
    n = a.length - 1
    steps_count = 0
    (0...n).to_a.each do |i|
      steps_count += shortest_path(a[i], b[i], a[i + 1], b[i + 1])
    end
    steps_count
  end

  private

  def shortest_path(ux, uy, vx, vy)
    dx = (ux - vx).abs
    dy = (uy - vy).abs
    [dx, dy].max
  end
end

File: Arrays/Min_Steps_in_Infinite_Grid/min_steps_in_infinite_grid.py
# https://www.interviewbit.com/problems/min-steps-in-infinite-grid

class Solution:
  # @param A : list of integers
  # @param B : list of integers
  # @return an integer
  def coverPoints(self, A, B):
    n = len(A)
    steps_count = 0
    for i in range(n - 1):
      steps_count += self.shortest_path(A[i], B[i], A[i + 1], B[i + 1])
    return steps_count
  
  def shortest_path(self, ux, uy, vx, vy):
    dx = abs(ux - vx)
    dy = abs(uy - vy)
    return max(dx, dy)
